plat/kvm/x86/traps.c

```diff
@@ -27,135 +27,198 @@
+		:
+		: "m"(gdtptr),
+		  "i"(GDT_DESC_OFFSET(GDT_DESC_CODE))
+		: "rax", "memory" : jump_to_new_cs);
+jump_to_new_cs:
 
-__section(".intrstack")  __align(STACK_SIZE)
-char cpu_intr_stack[STACK_SIZE];  /* IST1 */
-__section(".intrstack")  __align(STACK_SIZE)
-char cpu_trap_stack[STACK_SIZE];  /* IST2 */
-static char cpu_nmi_stack[4096];  /* IST3 */
+	__asm__ __volatile__(
+		/* Update remaining segment registers */
+		"movl	%0, %%es\n"
+		"movl	%0, %%ss\n"
+		"movl	%0, %%ds\n"
+
+		/* Initialize fs and gs to 0 */
+		"movl	%1, %%fs\n"
+		"movl	%1, %%gs\n"
+		:
+		: "r"(GDT_DESC_OFFSET(GDT_DESC_DATA)),
+		  "r"(0));
+}
 
-static void tss_init(void)
+static void tss_init(__lcpuidx idx)
 {
-	struct seg_desc64 *td = (void *) &cpu_gdt64[GDT_DESC_TSS_LO];
-
-	cpu_tss.ist[0] = (__u64) &cpu_intr_stack[sizeof(cpu_intr_stack)];
-	cpu_tss.ist[1] = (__u64) &cpu_trap_stack[sizeof(cpu_trap_stack)];
-	cpu_tss.ist[2] = (__u64) &cpu_nmi_stack[sizeof(cpu_nmi_stack)];
-
-	td->limit_lo = sizeof(cpu_tss);
-	td->base_lo = (__u64) &cpu_tss;
-	td->type = 0x9;
-	td->zero = 0;
-	td->dpl = 0;
-	td->p = 1;
-	td->limit_hi = 0;
-	td->gran = 0;
-	td->base_hi = (__u64) &cpu_tss >> 24;
-	td->zero1 = 0;
-
-	barrier();
+	struct seg_desc64 *tss_desc;
+
+	cpu_tss[idx].ist[0] =
+		(__u64) &cpu_intr_stack[idx][sizeof(cpu_intr_stack[idx])];
+	cpu_tss[idx].ist[1] =
+		(__u64) &cpu_trap_stack[idx][sizeof(cpu_trap_stack[idx])];
+
+	tss_desc = (void *) &cpu_gdt64[idx][GDT_DESC_TSS_LO];
+	tss_desc->limit_lo	= sizeof(cpu_tss[idx]);
+	tss_desc->base_lo	= (__u64) &(cpu_tss[idx]);
+	tss_desc->base_hi	= (__u64) &(cpu_tss[idx]) >> 24;
+	tss_desc->type		= 0x9; /* 64-bit TSS */
+	tss_desc->p		= 1;
+
 	__asm__ __volatile__(
-		"ltr %0"
+		"ltr %0\n"
 		:
-		: "r" ((unsigned short) (GDT_DESC_TSS_LO * 8))
-	);
+		: "r"((__u16) (GDT_DESC_OFFSET(GDT_DESC_TSS_LO))));
 }
 
-
 /* Declare the traps used only by this platform: */
 DECLARE_TRAP_EC(nmi,           "NMI",                  NULL)
 DECLARE_TRAP_EC(double_fault,  "double fault",         NULL)
 DECLARE_TRAP_EC(virt_error,    "virtualization error", NULL)
 
+static struct seg_gate_desc64 cpu_idt[IDT_NUM_ENTRIES] __align(8);
+static struct desc_table_ptr64 idtptr;
 
-static struct seg_gate_desc64 cpu_idt[IDT_NUM_ENTRIES] __align64b;
-
-static void idt_fillgate(unsigned int num, void *fun, unsigned int ist)
+static inline void idt_fillgate(unsigned int num, void *fun, unsigned int ist)
 {
 	struct seg_gate_desc64 *desc = &cpu_idt[num];
 
 	/*
 	 * All gates are interrupt gates, all handlers run with interrupts off.
 	 */
-	desc->offset_hi = (__u64) fun >> 16;
-	desc->offset_lo = (__u64) fun & 0xffff;
-	desc->selector = GDT_DESC_OFFSET(GDT_DESC_CODE);
-	desc->ist = ist;
-	desc->type = 14; /* == 0b1110 */
-	desc->dpl = 0;
-	desc->p = 1;
+	desc->offset_hi	= (__u64) fun >> 16;
+	desc->offset_lo	= (__u64) fun & 0xffff;
+	desc->selector	= GDT_DESC_OFFSET(GDT_DESC_CODE);
+	desc->ist	= ist;
+	desc->type	= 14; /* == 0b1110 */
+	desc->dpl	= 0;
+	desc->p		= 1;
```

Comment:
Define the macros for descriptor.

Reply:
Hi @Sairajkodilkar thanks for looking into this! Your help is very much appreciated! 😃

What do you mean with:
> Define the macros for descriptor.

Reply:
you can define the macros for desc->type, desc->dpl, desc->p

Reply:
> you can define the macros for descriptor type, privilege level.

Can you provide a suggestion on how you would like this to look?

Reply:
For example the IDT types can be defined as:
`#define IDT_GATE_TYPE_TASK 0x5`
`#define IDT_GATE_TYPE_INTR  0x6`
`#define IDT_GATE_TYPE_TRAP 0x7`